File: funtime/pylib/sokoban_draw.py
"""sokoban_draw.py - everything you can see in Sokoban, in Python."""

import math

from sokoban_rules import (LEVEL_WIDTH, LEVEL_HEIGHT, is_wall, is_goal)

CELL = 34
BOARD_MARGIN = 8

COLOR_INK = "#111111"
COLOR_PAPER = "#ffffff"
COLOR_FAINT = "#dddddd"


def board_pixel_width():
    """How wide the canvas must be."""
    return BOARD_MARGIN * 2 + LEVEL_WIDTH * CELL


def board_pixel_height():
    """How tall the canvas must be."""
    return BOARD_MARGIN * 2 + LEVEL_HEIGHT * CELL


def cell_left(x):
    """The left edge of one square."""
    return BOARD_MARGIN + x * CELL


def cell_top(y):
    """The top edge of one square."""
    return BOARD_MARGIN + y * CELL


def cell_at_pixel(x, y):
    """Which square did the player click? (x, y), or None."""
    column = int((x - BOARD_MARGIN) // CELL)
    row = int((y - BOARD_MARGIN) // CELL)
    if not (0 <= column < LEVEL_WIDTH) or not (0 <= row < LEVEL_HEIGHT):
        return None
    return (column, row)


def clear_canvas(ctx, width, height, color):
    """Paint the whole canvas one flat colour."""
    ctx.fillStyle = color
    ctx.fillRect(0, 0, width, height)


def draw_wall(ctx, x, y):
    """A hatched block."""
    left = cell_left(x)
    top = cell_top(y)
    ctx.fillStyle = COLOR_PAPER
    ctx.fillRect(left, top, CELL, CELL)
    ctx.strokeStyle = COLOR_INK
    ctx.lineWidth = 1
    i = -CELL
    while i < CELL:
        ctx.beginPath()
        ctx.moveTo(left + i, top + CELL)
        ctx.lineTo(left + i + CELL, top)
        ctx.stroke()
        i += 6
    ctx.lineWidth = 2
    ctx.strokeRect(left + 1, top + 1, CELL - 2, CELL - 2)


def draw_floor(ctx, x, y):
    """A plain square with a faint outline."""
    ctx.fillStyle = COLOR_PAPER
    ctx.fillRect(cell_left(x), cell_top(y), CELL, CELL)
    ctx.strokeStyle = COLOR_FAINT
    ctx.lineWidth = 1
    ctx.strokeRect(cell_left(x) + 0.5, cell_top(y) + 0.5, CELL - 1, CELL - 1)


def draw_goal(ctx, x, y):
    """The ring showing where a box belongs."""
    ctx.strokeStyle = COLOR_INK
    ctx.lineWidth = 2
    ctx.beginPath()
    ctx.arc(cell_left(x) + CELL / 2, cell_top(y) + CELL / 2, 6, 0, math.pi * 2)
    ctx.stroke()


def draw_box(ctx, x, y, home):
    """A crate: outlined, or solid black once it is home."""
    left = cell_left(x) + 4
    top = cell_top(y) + 4
    size = CELL - 8

    if home:
        ctx.fillStyle = COLOR_INK
        ctx.fillRect(left, top, size, size)
        ctx.strokeStyle = COLOR_PAPER
    else:
        ctx.fillStyle = COLOR_PAPER
        ctx.fillRect(left, top, size, size)
        ctx.strokeStyle = COLOR_INK
    ctx.lineWidth = 2
    ctx.strokeRect(left + 1, top + 1, size - 2, size - 2)
    ctx.beginPath()
    ctx.moveTo(left + 4, top + 4)
    ctx.lineTo(left + size - 4, top + size - 4)
    ctx.moveTo(left + size - 4, top + 4)
    ctx.lineTo(left + 4, top + size - 4)
    ctx.stroke()


def draw_player(ctx, player):
    """A little person: a head, a body and two feet."""
    middle_x = cell_left(player["x"]) + CELL / 2
    top = cell_top(player["y"])

    ctx.fillStyle = COLOR_INK
    ctx.beginPath()
    ctx.arc(middle_x, top + 11, 5.5, 0, math.pi * 2)
    ctx.fill()
    ctx.fillRect(middle_x - 5, top + 17, 10, 9)
    ctx.fillRect(middle_x - 7, top + 26, 4, 4)
    ctx.fillRect(middle_x + 3, top + 26, 4, 4)

    ctx.fillStyle = COLOR_PAPER
    ctx.fillRect(middle_x - 2, top + 8, 2, 2)
    ctx.fillRect(middle_x + 1, top + 8, 2, 2)


def draw_message(ctx, width, height, title, subtitle):
    """Big centred words across the level."""
    ctx.fillStyle = "rgba(255, 255, 255, 0.93)"
    ctx.fillRect(0, height / 2 - 40, width, 80)
    ctx.fillStyle = COLOR_INK
    ctx.textAlign = "center"

    scale = width / board_pixel_width()
    title_size = max(14, round(26 * scale))
    subtitle_size = max(9, round(12 * scale))

    ctx.font = "bold %dpx monospace" % title_size
    ctx.fillText(title, width / 2, height / 2 - 2)
    ctx.font = "%dpx monospace" % subtitle_size
    ctx.fillText(subtitle, width / 2, height / 2 + title_size * 0.8)
    ctx.textAlign = "left"


def render_game(ctx, state):
    """Draw one complete frame."""
    clear_canvas(ctx, board_pixel_width(), board_pixel_height(), COLOR_PAPER)

    for y in range(LEVEL_HEIGHT):
        for x in range(LEVEL_WIDTH):
            if is_wall(state, x, y):
                draw_wall(ctx, x, y)
            else:
                draw_floor(ctx, x, y)
                if is_goal(state, x, y):
                    draw_goal(ctx, x, y)

    for box in state["boxes"]:
        draw_box(ctx, box["x"], box["y"], is_goal(state, box["x"], box["y"]))

    draw_player(ctx, state["player"])

    ctx.strokeStyle = COLOR_INK
    ctx.lineWidth = 2
    ctx.strokeRect(BOARD_MARGIN - 1, BOARD_MARGIN - 1,
                   LEVEL_WIDTH * CELL + 2, LEVEL_HEIGHT * CELL + 2)

    if state["is_solved"]:
        draw_message(ctx, board_pixel_width(), board_pixel_height(), "SOLVED!",
                     "%d moves - press N for the next one" % state["moves"])
    elif state["is_paused"]:
        draw_message(ctx, board_pixel_width(), board_pixel_height(),
                     "PAUSED", "Press P to carry on")

File: funtime/lib/sokoban-draw.js
/* ============================================================
   sokoban-draw.js — everything you can see in Sokoban

   Black and white: a wall is a hatched block, a target is a small ring, a box
   is an outlined crate, and a box that is home is filled in solid — so you can
   count how many are done without reading any numbers.
   ============================================================ */

const CELL = 34;
const BOARD_MARGIN = 8;

const COLOR_INK = '#111111';
const COLOR_PAPER = '#ffffff';
const COLOR_FAINT = '#dddddd';

/** boardPixelWidth / boardPixelHeight — how big the canvas must be. */
function boardPixelWidth() { return BOARD_MARGIN * 2 + LEVEL_WIDTH * CELL; }
function boardPixelHeight() { return BOARD_MARGIN * 2 + LEVEL_HEIGHT * CELL; }

/** cellLeft / cellTop — the top-left corner of one square. */
function cellLeft(x) { return BOARD_MARGIN + x * CELL; }
function cellTop(y) { return BOARD_MARGIN + y * CELL; }

/** cellAtPixel — which square did the player click? null if none. */
function cellAtPixel(x, y) {
    const column = Math.floor((x - BOARD_MARGIN) / CELL);
    const row = Math.floor((y - BOARD_MARGIN) / CELL);
    if (column < 0 || column >= LEVEL_WIDTH || row < 0 || row >= LEVEL_HEIGHT) {
        return null;
    }
    return { x: column, y: row };
}

/** clearCanvas — paint the whole canvas one flat colour. */
function clearCanvas(ctx, width, height, color) {
    ctx.fillStyle = color;
    ctx.fillRect(0, 0, width, height);
}

/** drawWall — a hatched block. */
function drawWall(ctx, x, y) {
    const left = cellLeft(x);
    const top = cellTop(y);
    ctx.fillStyle = COLOR_PAPER;
    ctx.fillRect(left, top, CELL, CELL);
    ctx.strokeStyle = COLOR_INK;
    ctx.lineWidth = 1;
    for (let i = -CELL; i < CELL; i += 6) {
        ctx.beginPath();
        ctx.moveTo(left + i, top + CELL);
        ctx.lineTo(left + i + CELL, top);
        ctx.stroke();
    }
    ctx.lineWidth = 2;
    ctx.strokeRect(left + 1, top + 1, CELL - 2, CELL - 2);
}

/** drawFloor — a plain square with a faint outline. */
function drawFloor(ctx, x, y) {
    ctx.fillStyle = COLOR_PAPER;
    ctx.fillRect(cellLeft(x), cellTop(y), CELL, CELL);
    ctx.strokeStyle = COLOR_FAINT;
    ctx.lineWidth = 1;
    ctx.strokeRect(cellLeft(x) + 0.5, cellTop(y) + 0.5, CELL - 1, CELL - 1);
}

/** drawGoal — the ring showing where a box belongs. */
function drawGoal(ctx, x, y) {
    ctx.strokeStyle = COLOR_INK;
    ctx.lineWidth = 2;
    ctx.beginPath();
    ctx.arc(cellLeft(x) + CELL / 2, cellTop(y) + CELL / 2, 6, 0, Math.PI * 2);
    ctx.stroke();
}

/**
 * drawBox — a crate.
 * ALGORITHM: an outlined square with a cross through it, or a solid black one
 *            once it is standing on a target. Solid means done.
 */
function drawBox(ctx, x, y, home) {
    const left = cellLeft(x) + 4;
    const top = cellTop(y) + 4;
    const size = CELL - 8;

    if (home) {
        ctx.fillStyle = COLOR_INK;
        ctx.fillRect(left, top, size, size);
        ctx.strokeStyle = COLOR_PAPER;
    } else {
        ctx.fillStyle = COLOR_PAPER;
        ctx.fillRect(left, top, size, size);
        ctx.strokeStyle = COLOR_INK;
    }
    ctx.lineWidth = 2;
    ctx.strokeRect(left + 1, top + 1, size - 2, size - 2);
    ctx.beginPath();
    ctx.moveTo(left + 4, top + 4);
    ctx.lineTo(left + size - 4, top + size - 4);
    ctx.moveTo(left + size - 4, top + 4);
    ctx.lineTo(left + 4, top + size - 4);
    ctx.stroke();
}

/** drawPlayer — a little person: a head, a body and two feet. */
function drawPlayer(ctx, player) {
    const middleX = cellLeft(player.x) + CELL / 2;
    const top = cellTop(player.y);

    ctx.fillStyle = COLOR_INK;
    ctx.beginPath();
    ctx.arc(middleX, top + 11, 5.5, 0, Math.PI * 2);
    ctx.fill();
    ctx.fillRect(middleX - 5, top + 17, 10, 9);
    ctx.fillRect(middleX - 7, top + 26, 4, 4);
    ctx.fillRect(middleX + 3, top + 26, 4, 4);

    ctx.fillStyle = COLOR_PAPER;
    ctx.fillRect(middleX - 2, top + 8, 2, 2);
    ctx.fillRect(middleX + 1, top + 8, 2, 2);
}

/** drawMessage — big centred words across the level. */
function drawMessage(ctx, width, height, title, subtitle) {
    ctx.fillStyle = 'rgba(255, 255, 255, 0.93)';
    ctx.fillRect(0, height / 2 - 40, width, 80);
    ctx.fillStyle = COLOR_INK;
    ctx.textAlign = 'center';

    const scale = width / boardPixelWidth();
    const titleSize = Math.max(14, Math.round(26 * scale));
    const subtitleSize = Math.max(9, Math.round(12 * scale));

    ctx.font = 'bold ' + titleSize + 'px monospace';
    ctx.fillText(title, width / 2, height / 2 - 2);
    ctx.font = subtitleSize + 'px monospace';
    ctx.fillText(subtitle, width / 2, height / 2 + titleSize * 0.8);
    ctx.textAlign = 'left';
}

/** renderGame — draw one complete frame. */
function renderGame(ctx, state) {
    clearCanvas(ctx, boardPixelWidth(), boardPixelHeight(), COLOR_PAPER);

    for (let y = 0; y < LEVEL_HEIGHT; y++) {
        for (let x = 0; x < LEVEL_WIDTH; x++) {
            if (isWall(state, x, y)) {
                drawWall(ctx, x, y);
            } else {
                drawFloor(ctx, x, y);
                if (isGoal(state, x, y)) {
                    drawGoal(ctx, x, y);
                }
            }
        }
    }

    for (let i = 0; i < state.boxes.length; i++) {
        const box = state.boxes[i];
        drawBox(ctx, box.x, box.y, isGoal(state, box.x, box.y));
    }

    drawPlayer(ctx, state.player);

    ctx.strokeStyle = COLOR_INK;
    ctx.lineWidth = 2;
    ctx.strokeRect(BOARD_MARGIN - 1, BOARD_MARGIN - 1,
                   LEVEL_WIDTH * CELL + 2, LEVEL_HEIGHT * CELL + 2);

    if (state.isSolved) {
        drawMessage(ctx, boardPixelWidth(), boardPixelHeight(), 'SOLVED!',
                    state.moves + ' moves — press N for the next one');
    } else if (state.isPaused) {
        drawMessage(ctx, boardPixelWidth(), boardPixelHeight(), 'PAUSED', 'Press P to carry on');
    }
}
